### PR TITLE
Add support for parsing GSUB/GPOS format 1.1

### PR DESCRIPTION
**Pull request by jfkthame**
In variation fonts, a new minor version of GSUB/GPOS may be used, to include a Feature Variations subtable. So this patch adds support for parsing that.

### CODE REVIEW

File: src/layout.cc [hunk 1]
```diff
@@ -1496,6 +1496,160 @@ bool ParseExtensionSubtable(const Font *font,
...
+  for (uint16_t i = 0; i < substitution_count; i++) {
+    uint16_t feature_index = 0;
+    uint32_t alternate_feature_table_offset = 0;
+
+    if (!subtable.ReadU16(&feature_index) ||
+        !subtable.ReadU32(&alternate_feature_table_offset)) {
+      return OTS_FAILURE_MSG("Failed to read feature table substitution record");
+    }
+
+    if (alternate_feature_table_offset < 6 + 6 * substitution_count ||
```
Comment (khaledhosny) [inline src/layout.cc]: Should these magic numbers get names like `kFeatureTableSubstitutionHeaderSize` and `kFeatureTableSubstitutionRecordSize` for readability?

File: src/layout.cc [hunk 2]
```diff
@@ -1496,6 +1496,160 @@ bool ParseExtensionSubtable(const Font *font,
...
+
+  for (uint32_t i = 0; i < feature_variation_record_count; i++) {
+    uint32_t condition_set_offset = 0;
+    uint32_t feature_table_substitution_offset = 0;
+    if (!subtable.ReadU32(&condition_set_offset) ||
+        !subtable.ReadU32(&feature_table_substitution_offset)) {
+      return OTS_FAILURE_MSG("Failed to read feature variation record");
+    }
+
+    if (condition_set_offset < subtable.offset() ||
```
Comment (khaledhosny) [inline src/layout.cc]: Shouldn’t this be `subtable.offset() + 8` to account for the header?

File: src/layout.cc (hunk 1 above)
Comment (khaledhosny) [reply src/layout.cc]: Also, shouldn’t this be `subtable.offset() + …?`

File: src/layout.cc [hunk 3]
```diff
@@ -1496,6 +1496,160 @@ bool ParseExtensionSubtable(const Font *font,
...
+        condition_set_offset > length ||
+        feature_table_substitution_offset < subtable.offset() ||
```
Comment (khaledhosny) [inline src/layout.cc]: Same here.

File: src/layout.cc [hunk 4]
```diff
@@ -1496,6 +1496,160 @@ bool ParseExtensionSubtable(const Font *font,
...
+        feature_table_substitution_offset > length) {
+      return OTS_FAILURE_MSG("Offset out of range");
+    }
+
+    if (condition_set_offset) {
```
Comment (khaledhosny) [inline src/layout.cc]: Shouldn’t this offset be validated as well?

File: src/layout.cc [hunk 5]
```diff
@@ -1496,6 +1496,160 @@ bool ParseExtensionSubtable(const Font *font,
...
+      if (!ParseConditionSetTable(font, data + condition_set_offset,
+                                  length - condition_set_offset)) {
+        return OTS_FAILURE_MSG("Failed to parse condition set table");
+      }
+    }
+
+    if (feature_table_substitution_offset) {
```
Comment (khaledhosny) [inline src/layout.cc]: And this one?

**Comment by khaledhosny** [comment]
The ci failures should go if you rebase on master.

File: src/layout.cc (hunk 1 above)
Comment (jfkthame) [reply src/layout.cc]: > Should these magic numbers get names...
Sure, sounds good.

> Also, shouldn’t this be `subtable.offset() + …`?
I don't think so; this is an offset from the start of the FeatureTableSubstitution table, not from the current offset within it.

**Comment by jfkthame** [comment]
Rebased, fixed up offset validations, and squashed to make a tidier commit; let's see how this looks.

**Comment by jfkthame** [comment]
AdobeVFPrototype (license: OFL) uses table version 1.1, so sanitization fails without the patch here (but works once it is applied).

**Pull request merged**